### PR TITLE
R3

### PR DESCRIPTION
**Pull request by qubotong**
@fwilliams 
This is the version created to generate the scene. 
Some scenes have been tested and look correct. 
The memory may running out when writing the scenes with too many fds.

### CODE REVIEW

File: visualiser/skybox.py [hunk 1]
```diff
@@ -2,9 +2,10 @@
 
 
 class Skybox(object):
-    def __init__(self, bottom_color=(0, 0, 0), top_color=(0, 0, 0)):
+    def __init__(self, bottom_color=(0, 0, 0), top_color=(0, 0, 0), surrounding_color=(0.5, 0.0, 0.0)):
```
Comment (fwilliams) [inline visualiser/skybox.py]: What is `surrounding_color`?  Why do we need it?

File: visualiser/viewer.py [hunk 2]
```diff
@@ -12,18 +15,152 @@
...
+        self._showNormals = False
+        self._showWires = False
+        self._showSamples = False
+
+    def initUI(self):
+        self.exitAction = QAction('&Exit', self)
+        self.exitAction.setToolTip('Exit Application')
+        self.exitAction.triggered.connect(qApp.quit)
+
+        self.aboutAction = QAction('&Help', self)
```
Comment (fwilliams) [inline visualiser/viewer.py]: Do we really need an about Dialog? It seems excessive.

File: visualiser/viewer.py [hunk 3]
```diff
@@ -12,18 +15,152 @@
...
+
+        self.toolbar = self.addToolBar('ToolBar')
+        self.toolbar.addAction(self.aboutAction)
+        self.toolbar.addAction(self.exitAction)
+        self.toolbar.addAction(self.toggleShowWireAction)
+        self.toolbar.addAction(self.toggleShowSampleAction)
+        self.toolbar.addAction(self.toggleShowNormalsAction)
+        self.toolbar.addAction(self.toggleShowAxesAction)
+
+        self.statusBar().showMessage('Ready') #only works for QMainWindow
```
Comment (fwilliams) [inline visualiser/viewer.py]: What is this doing?

File: visualiser/viewer.py [hunk 4]
```diff
@@ -12,18 +15,152 @@
...
+
+    @property
+    def showWires(self):
+        return self._showWires
+
+    @property
+    def showSamples(self):
+        return self._showSamples
+
+class AboutDlg(QWidget):
```
Comment (fwilliams) [inline visualiser/viewer.py]: Again, this seems like extra code for no reason. If you want to have instructions, add it to the `README.md`

File: geometry/shapes.py [hunk 5]
```diff
@@ -432,3 +432,45 @@ def intersects(obj1, obj2):
...
+
+
+class Triangle(Polyhedron):
+    def __init__(self, *args):
+        super(Triangle, self).__init__()
+
+        self._vertices = []
+        input_shape = np.array(args[0]).shape
+
+        if len(args) != 3:
```
Comment (fwilliams) [inline geometry/shapes.py:420]: If I pass in a 3x3 array, you won't catch an error. You need to explicitly check that it's a 3x1 or 1x3.

File: geometry/shapes.py [hunk 6]
```diff
@@ -432,3 +432,45 @@ def intersects(obj1, obj2):
...
+                raise ValueError("Triangle base vertices do not all have the same dimension")
+
+            if pt.shape == (2,):
+                pt = np.array((pt[0], pt[1], 0.0, 1.0))
+            elif pt.shape == (3,):
+                pt = np.array((pt[0], pt[1], pt[2], 1.0))
+            elif pt.shape == (4,):
+                pt = utils.make_projective_point(pt)
+            else:
+                raise ValueError("Prism cannot be constructed from non 2d, 3d or projective 3d points. Got %s" % pt)
```
Comment (fwilliams) [inline geometry/shapes.py]: Change the error text here. This is not a prism class. Actually, it would be super cool to move this logic to a function in `utils.py` since we seem to be doing it everywhere. Why not make it a function and update the other places in the code too?

File: main.py [hunk 7]
```diff
@@ -49,32 +47,62 @@ def init(viewer):
...
+
+    glPopAttrib(GL_ENABLE_BIT)
+    glEndList()
+
+    global sample_display_list
+    sample_display_list = glGenLists(1)
+    glNewList(sample_display_list, GL_COMPILE)
+    glPushAttrib(GL_ENABLE_BIT)
+
+    for t in tilemap.values():
```
Comment (fwilliams) [inline main.py:100]: What exactly is this drawing? A triangle per fundamental domain? Why?

File: main.py [hunk 8]
```diff
@@ -123,11 +166,41 @@ def resize(viewer):
...
+# args.radius = 15
+# args.vradius = 10
+# args.overlap = 0
+# args.scale = 560
+# args.height = 560;
+# args.bidir = False;
+# args.visualize = True;
+# args.ceiling = True;
+# args.floor = True;
+
```
Comment (fwilliams) [inline main.py]: Get rid of this if it's just commented out.

File: main.py [hunk 9]
```diff
@@ -164,30 +237,58 @@ def resize(viewer):
...
+        vbase_kernel = tiling.vLineKernel(args.vradius, (0,0), vgroup, base_kernel, args.ceiling, args.floor)
+
+    elif args.type == "x2222" or args.type == "x442":
+        vbase_kernel = tiling.vSquareKernel(args.vradius, (0,0,0), vgroup, base_kernel, args.ceiling, args.floor)
+
+    elif args.type == "x632" or args.type == "x333":
+        vbase_kernel = tiling.vHexKernel(args.vradius, (0, 0, 0), vgroup, base_kernel, args.ceiling, args.floor)
+
+    kt = tiling.KernelTiling(vbase_kernel, frustum, args.overlap)
+
```
Comment (fwilliams) [inline main.py]: This looks like duplicated code. In lines 202 to 240 we have a similar switch. Why do we need this? I don't understand what it's doing?

File: geometry/shapes.py (hunk 6 above)
Comment (fwilliams) [inline geometry/shapes.py]: Change the error message here. It's a triangle, not a prism.

Also, since it looks like this conversion from 2d to 3d is happening everywhere, why not make it a utility function in `utils.py`? Then update the code everywhere to use it. It'll be much less of a pain in the ass to read. 

In general, don't copy paste code. If you're doing something twice or more, make it a function.

File: main.py (hunk 9 above)
Comment (fwilliams) [inline main.py]: This is duplicated logic from lines 202 through 240 of this file. See my comment on `tiling.py` for more info.

File: tiling.py [hunk 10]
```diff
@@ -21,10 +21,11 @@ def __init__(self, height, up, mirror1_ctr, mirror2_ctr):
         up_proj = up - np.dot(up, m1n)*m1n
         up_proj /= np.linalg.norm(up_proj)
         right_proj = utils.make_projective_vector(np.cross(up_proj[0:3], m1n[0:3]))
+        right_proj2 = utils.make_projective_vector(np.cross(up_proj[0:3], m2n[0:3]))
```
Comment (fwilliams) [inline tiling.py:23]: Why do you need this? What's happening here?

File: tiling.py [hunk 11]
```diff
@@ -310,6 +315,119 @@ def fundamental_domain_transforms(self):
...
+
+    @property
+    def radius(self):
+        return self._radius
+
+    @property
+    def group(self):
+        return self._group
+
+class vSquareKernel(object):
```
Comment (fwilliams) [inline tiling.py]: Let's not copy-paste classes to make small modifications. Why not just simply augment the existing square kernel class to handle 3D cases. It should be 5 lines of code to do this...

File: tiling.py [hunk 12]
```diff
@@ -368,6 +486,177 @@ def fundamental_domain_transforms(self):
...
+    def radius(self):
+        return self._radius
+
+    @property
+    def group(self):
+        return self._group
+
+
+
+class vHexKernel(object):
```
Comment (fwilliams) [inline tiling.py]: Same comment as above with `vSquareKernel`. This is not maintainable at all and hard to follow.

File: tiling.py [hunk 13]
```diff
@@ -401,7 +690,7 @@ def translational_fundamental_domains(self):
         for i in range(self._diameter):
             pos = np.array((i - self._radius)) + np.array(self._center)
             translate = utils.translation_matrix(
-                pos * self._group.translational_subgroup_basis[1 if pos > 0 else 0])
+                abs(pos) * self._group.translational_subgroup_basis[1 if pos > 0 else 0])
```
Comment (fwilliams) [inline tiling.py]: What is the abs fixing?

File: tiling.py [hunk 14]
```diff
@@ -422,29 +711,103 @@ def fundamental_domain_transforms(self):
...
+    @property
+    def radius(self):
+        return self._radius
+
+    @property
+    def group(self):
+        return self._group
+
+
+class vLineKernel(object):
```
Comment (fwilliams) [inline tiling.py]: Again, please don't duplicate entire classes, just add the feature to what's already here. I expect this file should not have to change very much.

File: tiling.py [hunk 15]
```diff
@@ -368,6 +486,177 @@ def fundamental_domain_transforms(self):
     def center(self):
         return self._center
 
+    @property
+    def diameter(self):
```
Comment (fwilliams) [inline tiling.py]: Why do we need diameter, radius and group properties?

File: tiling.py [hunk 16]
```diff
@@ -422,29 +711,103 @@ def fundamental_domain_transforms(self):
...
     def fundamental_domains(self):
-        for k in range(len(self._group.dihedral_subgroup)):
-            transform = self._group.dihedral_subgroup[k]
-            prism = shapes.Prism(self._group.height, *self._group.fd_vertices)
-            prism.transform(transform)
-            yield (0, k), transform, prism
+        for pos, translate, _ in self.translational_fundamental_domains:
+            for k in range(len(self._vgroup.dihedral_subgroup)):
+
+                for k_r2 in range(len(self._r2_base_kernel.group.dihedral_subgroup)):
```
Comment (fwilliams) [inline tiling.py]: What's going on here? You're trying to handle the floor and ceiling cases right? This seems way overcomplicated. Here's some quick pseudocode of how I would do this. 

```
yield (0, k, 0), transform, prism
if ceiling:
    yield (0, k, 1), transform*ceiling_reflection, transform(prism, ceiling_reflection)
elif floor:
    yield (0, k, 1), transform*floor_reflection, transform(prism, floor_reflection)
```

Comment (fwilliams) [reply tiling.py]: Also, like I said, just modify the existing code.

File: tiling.py [hunk 17]
```diff
@@ -422,29 +711,103 @@ def fundamental_domain_transforms(self):
...
+                    prism = shapes.Prism(self._vgroup.height, *self._r2_base_kernel.group.fd_vertices)
+                    prism.transform(transform)
+                    yield (pos, k_r2, k), transform, prism
 
     @property
     def translational_fundamental_domains(self):
-        prism = shapes.Prism(self._group.height, *self._group.translational_fd_vertices)
-        yield 0, np.eye(4), prism
+
+        for i in range(self._diameter):
```
Comment (fwilliams) [inline tiling.py]: Again, this turned 2 lines of code into 30. Why? Just update the existing kernel to handle this nicely. It should be a few lines of code to change.

File: tiling.py [hunk 18]
```diff
@@ -458,8 +821,46 @@ def fundamental_domain_transforms(self):
...
+#
+#     @property
+#     def fundamental_domain_transforms(self):
+#         for _, tx, _ in self.fundamental_domains:
+#             yield tx
+#
+#     @property
+#     def center(self):
+#         return 0
+#
```
Comment (fwilliams) [inline tiling.py:646]: Don't commit commented out blocks like this.

File: visualiser/viewer.py [hunk 19]
```diff
@@ -12,18 +15,152 @@
...
+#         super(Window, self).__init__()
+#
+#         self.gl_widget = GLWidget()
+#
+#         main_layout = QtWidgets.QHBoxLayout()
+#         main_layout.addWidget(self.gl_widget)
+#
+#         self.setLayout(main_layout)
+#         self.setWindowTitle("Hello GL")
+
```
Comment (fwilliams) [inline visualiser/viewer.py]: Kill commented blocks.

File: visualiser/viewer.py [hunk 20]
```diff
@@ -97,11 +234,14 @@ def __init__(self, title="GL Viewer", size=(800, 600)):
...
+        self._window = Window()
+        self._window.setGLWidget(self.gl_widget)
 
-        self.setLayout(main_layout)
-        self.setWindowTitle(title)
+        # main_layout = QtWidgets.QHBoxLayout()
+        # main_layout.addWidget(self.gl_widget)
+        #
+        # self.setLayout(main_layout)
+        # self.setWindowTitle(title)
```
Comment (fwilliams) [inline visualiser/viewer.py]: Kill commented blocks.

**Review by fwilliams — CHANGES_REQUESTED** [review]
Overall, your images look good but the code is really hard to follow which means that finding bugs and updating it is going to be challenging. I added a bunch of comments, please fix them and update the pull request.

File: main.py (hunk 7 above)
Comment (qubotong) [reply main.py:100]: These samples are colorful triangles inside each fd. This is really helpful for checking whether the reflections are correct.

File: tiling.py (hunk 18 above)
Comment (qubotong) [reply tiling.py:646]: I kind of feel this part of your old codes was not useful anymore. But not sure if you saved them for future development. So I just commented them.

File: geometry/shapes.py (hunk 5 above)
Comment (qubotong) [reply geometry/shapes.py:420]: Each arg is checked if it is legal by a new added function utils.make_legal_point(pt, input_shape)

File: geometry/shapes.py (hunk 6 above)
Comment (qubotong) [reply geometry/shapes.py]: a new function utils.make_legal_point(pt, input_shape) is added

Comment (qubotong) [reply geometry/shapes.py]: done

File: main.py (hunk 8 above)
Comment (qubotong) [reply main.py]: I still need them for future development. I will remove it at last.

**Review by qubotong — COMMENTED** [review]
1. The writing function has been rewritten to make sure the writing process will not use lots memory to save the nodes
2. The kernel functions are changed to be able to handle both the horizontal and vertical fds.

We still need to merge some nodes together when designing the scene. The pumpkin 632 opath scene can be 3.2 GB which I assume is too big.

File: main.py [hunk 21]
```diff
@@ -109,45 +151,83 @@ def draw(viewer):
...
+# args.hradius = 20
+# args.vradius = 10
+# args.overlap = 1
+# args.scale = 560
+# args.height = 1000;
+# args.bidir = False;
+# args.visualize = True;
+# args.ceiling = False;
+# args.floor = False;
+# args.inc = True;
```
Comment (fwilliams) [inline main.py:200]: Kill commented out code.

File: tiling.py [hunk 22]
```diff
@@ -255,45 +260,106 @@ def mirror_planes(self):
...
+                        yield (pos, k_h, k_v), transform, prism
 
     @property
     def translational_fundamental_domains(self):
-        for i in range(self._diameter):
-            for j in range(self._diameter):
-                pos = np.array((i - self._radius, j - self._radius)) + np.array(self._center)
-                translate = utils.translation_matrix(
-                    pos[0] * self._group.translational_subgroup_basis[0] +
-                    pos[1] * self._group.translational_subgroup_basis[1])
```
Comment (fwilliams) [inline tiling.py:291]:
```
....
for j in range(self._vdiameter):
    translate = utils.translation_matrix(		
                    pos[0] * self._group.translational_subgroup_basis[0] +		
                    pos[1] * self._group.translational_subgroup_basis[1]) + 
                    pos[2]*self._group._translational_subgroup_basis[2]
```

**Comment by qubotong** [comment]
Will start new prs with better structured codes
